### ISSUE TITLE
d8 does not run on G4 Mac

### ISSUE THREAD

**Issue by spotlightishere** [andrewlow/v8ppc#120]
Yes, I've read lots of the other G4 related problems. But, I don't think this is related.
- I believe this was compiled using clang
- Snapshot was off (it doesn't work anyway)
- I ran GDB on it and got this:

```
(gdb) run
Starting program: /home/me/g4compat/out/ppc.release/d8 
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/powerpc-linux-gnu/libthread_db.so.1".
[New Thread 0x4883d480 (LWP 28348)]

Program received signal SIGILL, Illegal instruction.
0x5185ebe8 in ?? ()
(gdb) x/20i $pc
=> 0x5185ebe8:  fcfid   f1,f1
   0x5185ebec:  clrlwi. r0,r5,31
   0x5185ebf0:  srawi   r11,r5,1
   0x5185ebf4:  beq     0x5185ed70
   0x5185ebf8:  lwz     r11,-1(r5)
   0x5185ebfc:  cmpw    cr7,r11,r8
   0x5185ec00:  bne     cr7,0x5185ee1c
   0x5185ec04:  lfd     f2,3(r5)
   0x5185ec08:  fctidz  f0,f2
   0x5185ec0c:  addi    r1,r1,-8
   0x5185ec10:  stfd    f0,0(r1)
   0x5185ec14:  ori     r2,r2,0
   0x5185ec18:  lwz     r10,0(r1)
   0x5185ec1c:  lwz     r11,4(r1)
   0x5185ec20:  addi    r1,r1,8
   0x5185ec24:  srawi   r0,r11,31
   0x5185ec28:  cmpw    cr7,r0,r10
   0x5185ec2c:  bne     cr7,0x5185ec38
   0x5185ec30:  fcfid   f0,f0
   0x5185ec34:  fcmpu   cr7,f0,f2
(gdb) 
```

If you need other info, please ask.

**Comment by spotlightishere** [comment]
I'm not on OS X right now (linux is faster) but with Xcode 3.2, they have an instruction reference book. I don't think it's legal to share it but it has some important tips and usage things.

**Comment by andrewlow** [comment]
Caught me on a holiday break from the computer -- @tunniclm had done some hacking here and gotten most of the way to making it work on the G4 (under Linux). He captured most of the work in this branch https://github.com/andrewlow/v8ppc/tree/g4compat .. probably out of date at this point, but should have been in a working state.

**Comment by awesomebing1** [comment]
@andrewlow I did compile that from the g4compat branch, and it has an illegal instruction for me.

**Comment by tunniclm** [comment]
@awesomebing1 Hi, fcfid is one of the instructions that the g4compat branch should be avoiding. It does attempt to detect incompatible processors, so it may be that the detection is not triggering.

Can you run `LD_SHOW_AUXV=1 /bin/true` to check the processor name? Currently the code is quite restricted and only checks for specific known processors, and will assume it has found a G4 if the AT_PLATFORM is 'ppc7450'.

**Comment by awesomebing1** [comment]
That's a pretty cool command. Yes, the platform is ppc7450.


If it was compiled with clang and snapshot off, would that make a difference?

**Comment by andrewlow** [comment]
Next step would be to try to identify WHERE this fcfid is being emitted from.

You could put a breakpoint into the src/ppc/assembler-ppc.cc code where it emits the instruction. I'm actually a bit suspicious of clang possibly sneaking it in, do you have proof that clang works ok?

**Comment by tunniclm** [comment]
0x5185ec08:  fctidz  f0,f2
    0x5185ec0c:  addi    r1,r1,-8
    0x5185ec10:  stfd    f0,0(r1)
    0x5185ec14:  ori     r2,r2,0
    0x5185ec18:  lwz     r10,0(r1)
    0x5185ec1c:  lwz     r11,4(r1)
    0x5185ec20:  addi    r1,r1,8
    0x5185ec24:  srawi   r0,r11,31
    0x5185ec28:  cmpw    cr7,r0,r10
    0x5185ec2c:  bne     cr7,0x5185ec38
    0x5185ec30:  fcfid   f0,f0
    0x5185ec34:  fcmpu   cr7,f0,f2

That chunk looks a lot like something that would be emitted by EmitVFPTruncate() [macro-assembler-ppc.cc:2370] if the call to  EmitVFPTruncate_NoPPC64() was not taken.

I can't recognise where the start might come from though.

**Comment by awesomebing1** [comment]
Yes. My clang does work. I have not encountered any errors. It's clang 3.5 compiled with clang 3, on Debian. I can add a breakpoint later.

**Comment by andrewlow** [comment]
Weird - I just checked out the g4compat branch.

```
$ cd v8ppc/src/ppc
$ grep fcfid *
assembler-ppc.cc:void Assembler::fcfid(const DwVfpRegister frt,
assembler-ppc.h:  void fcfid(const DwVfpRegister frt, const DwVfpRegister frb,
code-stubs-ppc.cc:  __ fcfid(double_dst, double_dst);
disasm-ppc.cc:      Format(instr, "fcfid'.  'Dt, 'Db");
macro-assembler-ppc.cc:    fcfid(double_scratch, double_scratch);

```

So code-stubs-ppc.cc and macro-assembler-ppc.cc would be the two locations where I'd expect to see this instruction.

code-stubs-ppc.cc
```
  // Convert to double word FP from stack
  __ lfd(double_dst, MemOperand(sp, 0));
  __ fcfid(double_dst, double_dst);

  __ addi(sp, sp, Operand(8));  // restore stack
```

macro-assembler-ppc.cc
```
    // convert back and compare
    fcfid(double_scratch, double_scratch);
    fcmpu(double_scratch, double_input);
    bind(&done);
```

Neither of them seem to be the sequence that matches the bad one.

```
=> 0x5185ebe8:  fcfid   f1,f1
   0x5185ebec:  clrlwi. r0,r5,31
   0x5185ebf0:  srawi   r11,r5,1
   0x5185ebf4:  beq     0x5185ed70
   0x5185ebf8:  lwz     r11,-1(r5)
```

Feels like I'm missing something here?

**Comment by barracuda156** [comment]
So does it work on MacOS PowerPC in the current state or not? I need V8 for some R functionality.
